### PR TITLE
HTML-705: HTML Form Entry tag for creating a single condition

### PR DESCRIPTION
**Pull request by samuelmale**
Ticket : https://issues.openmrs.org/browse/HTML-705

### CODE REVIEW

File: api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+			if (givenOnsetDate != null && givenEndDate != null) {
+				if (givenOnsetDate.after(givenEndDate)) {
+					ret.add(new FormSubmissionError(context
+	                        .getFieldName(endDate), "htmlformentryui.conditionui.endDate.before.onsetDate.error"));
+				}
+			} 
+			if (status == null && (required || StringUtils.isNotBlank(condition))) {
+				ret.add(new FormSubmissionError(context
+	                    .getFieldName(conditionStatusWidget), "htmlformentryui.conditionui.status.required"));
+			}
```
Comment (mks-d) [inline api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java]: I'm confused now, I thought all those validations were ensured at the service layer level?

Comment (samuelmale) [reply api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java]: I think this is required before the form is submitted!

- It prevents the form from being submitted once a condition is required and one isn't given ie. `<condition required="true" />` .
- It prevents form submission when also one provided the Condition and didn't select a status. This applies to cases where the condition is option ie. `<condition required="false"/>`

Comment (mks-d) [reply api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java]: The case of `required` should be handled here for sure, i.e. making sure that the condition field is filled in if the tag is required.

>It prevents form submission when also one provided the Condition and didn't select a status.

But that case should be picked up by the service layer no?

Comment (samuelmale) [reply api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java]: Ok. I guess have fixed that?

File: api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+		ret.append("     var valueAttr = jq('#" + conditionNameTextInputId + "_hid" + "').attr('value');\n");
+		ret.append("     if(valueAttr === \"ERROR\"){\n");
+		ret.append("        jq('#" + conditionNameTextInputId + "_hid" + "').attr('value', '');\n");
+		ret.append("     }\n");
+		ret.append("});\n");				
+		ret.append("</script>\n");
+		return ret.toString();
+	}
+	
+	private String htmlForConditionStatusWidgets(FormEntryContext context) {
```
Comment (mks-d) [inline api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java]: A detail but rather: `htmlForConditionStatusesWidget`

File: api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+				"}");
+		sb.append("#condition-status {\n" +
+				 "   padding:10px 0px;\n" +
+				 " }");
+		sb.append("</style>");
+		sb.append("</div>");
+		return sb.toString();	
+	}
+	
+	private String htmlForConditionDateWidgets(FormEntryContext context) {
```
Comment (mks-d) [inline api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java]: A detail but rather: `htmlForConditionDatesWidget `

File: api-2.2/src/test/java/org/openmrs/htmlformentryui/element/ConditionElementTest.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+	@Mock
+	private ConditionService conditionService;
+	@Mock
+	private FormEntrySession session;
+	@Mock
+	private FormEntryContext context;
+	@Mock
+	private ConceptSearchAutocompleteWidget conditionSearchWidget;
+	@Mock
+	private RadioButtonsWidget conditionStatusWidget;
```
Comment (mks-d) [inline api-2.2/src/test/java/org/openmrs/htmlformentryui/element/ConditionElementTest.java]: `conditionStatusesWidget `

File: api-2.2/src/test/java/org/openmrs/htmlformentryui/element/ConditionElementTest.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+		// replay
+		element.handleSubmission(session, request);
+		
+		// verify
+		ArgumentCaptor<Condition> captor = ArgumentCaptor.forClass(Condition.class);
+		verify(conditionService).saveCondition(captor.capture());
+		Condition condition = captor.getValue();
+		Assert.assertEquals(ConditionClinicalStatus.ACTIVE, condition.getClinicalStatus());
+		Assert.assertNotNull(condition.getPatient());
+		Assert.assertNotNull(condition.getCondition());
```
Comment (mks-d) [inline api-2.2/src/test/java/org/openmrs/htmlformentryui/element/ConditionElementTest.java]: Is there no way here to verify that the condition is indeed based on concept 1519?

File: api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+
+public class ConditionElement implements HtmlGeneratorElement, FormSubmissionControllerAction  {
+	
+	private MessageSourceService mss;
+	private ConditionService conditionService;
+	private boolean required;
+	// widgets
+	private Widget conditionSearch;
+	private DateWidget onSetDate;
+	private DateWidget endDate;
```
Comment (mks-d) [inline api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java]: Suffix all widgets by `Widget`:
```java
private Widget conditionSearchWidget;
private DateWidget onSetDateWidget;
private DateWidget endDateWidget;
```

File: api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+		if (context.getMode() != Mode.VIEW) {
+			if (StringUtils.isBlank(condition) && required) {
+				ret.add(new FormSubmissionError(context.getFieldName(conditionSearch), 
+		                    Context.getMessageSourceService().getMessage("htmlformentryui.conditionui.condition.required")));
+				
+			}	
+			if (givenOnsetDate != null && givenEndDate != null) {
+				if (givenOnsetDate.after(givenEndDate)) {
+					ret.add(new FormSubmissionError(context.getFieldName(endDate), 
+							Context.getMessageSourceService().getMessage("htmlformentryui.conditionui.endDate.before.onsetDate.error")));
```
Comment (mks-d) [inline api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java]: Remove the extra blank line.

File: api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+				}
+			} 
+			if (status == null && required) {
+				ret.add(new FormSubmissionError(context.getFieldName(conditionStatusesWidget), 
+						Context.getMessageSourceService().getMessage("htmlformentryui.conditionui.status.required")));
```
Comment (mks-d) [inline api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java]: Remove the extra blank line.

File: api/src/main/resources/messages.properties [hunk 9]
```diff
@@ -10,4 +10,10 @@ htmlformentryui.saveForm = Save Form
...
 
-htmlformentryui.viewHtmlForm.breadcrumb=View: {0}
\ No newline at end of file
+htmlformentryui.viewHtmlForm.breadcrumb=View: {0}
+
+# Condition UI
+htmlformentryui.conditionui.historyOf.label=History Of
+htmlformentryui.conditionui.condition.required=Condition required
+htmlformentryui.conditionui.status.required=Condition status required
+htmlformentryui.conditionui.endDate.before.onsetDate.error=End date can't be before onset date
```
Comment (mks-d) [inline api/src/main/resources/messages.properties]: As close as possible to full written English please:
```java
End date cannot be before onset date
```

File: api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+		context.registerErrorWidget(conditionSearch, conditionSearchErrorWidget);
+		
+		StringBuilder ret = new StringBuilder();
+		ret.append(conditionSearch.generateHtml(context));
+		if (context.getMode() != Mode.VIEW) {
+			ret.append(conditionSearchErrorWidget.generateHtml(context));
+		}
+		ret.append("\n<script>jq('#" + conditionNameTextInputId + "').attr('placeholder',");
+		ret.append(" '" + mss.getMessage("coreapps.conditionui.condition") + "');\n");
+		ret.append(" jq('#" + conditionNameTextInputId + "').css('min-width', '46.4%');\n");
```
Comment (mks-d) [inline api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java]: @mogoodrich, @samuelmale did inline some widget-specific CSS here, is that ok? Or how do you handle this in general?

File: api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+		ret.append("</ul> <br/>");
+		ret.append("<script> jq('#" + onsetDateTextInputId + "').attr('placeholder',");
+		ret.append(" '" + mss.getMessage("coreapps.conditionui.onsetdate") + "');");
+		ret.append("jq('#" + endDateTextInputId + "').attr('placeholder',");
+		ret.append(" '" + mss.getMessage("coreapps.stopDate.label") + "');");
+		ret.append("</script>");
+		ret.append("<style>");
+		ret.append("#htmlformentryui-condition li {\n" + 
+				"	width:30%;\n" + 
+				"   float: left;\n" +
```
Comment (mks-d) [inline api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java]: @mogoodrich, here again.

File: api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+		ret.append(" '" + mss.getMessage("coreapps.stopDate.label") + "');");
+		ret.append("</script>");
+		ret.append("<style>");
+		ret.append("#htmlformentryui-condition li {\n" + 
+				"	width:30%;\n" + 
+				"   float: left;\n" + 
+				"}");
+		ret.append("#htmlformentryui-condition ul {\n" + 
+				"	display:flow-root;\n" + 
+				"   width:85%" +
```
Comment (mks-d) [inline api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java]: @mogoodrich, here again.

File: api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+		return ret.toString();
+	}
+		
+	private String htmlForConditionSearchWidget(FormEntryContext context) {
+		List<ConceptClass> requiredClasses = new ArrayList<ConceptClass>();
+		Set<Concept> initialConcepts = new HashSet<Concept>();
+		if (mss == null) {
+			mss = Context.getMessageSourceService();		
+		}
+		ConceptClass classCache = Context.getConceptService().getConceptClassByUuid("8d4918b0-c2cc-11de-8d13-0010c6dffd0f");
```
Comment (mks-d) [inline api-2.2/src/main/java/org/openmrs/htmlformentryui/ConditionElement.java]: `classCache` → `conceptClass`

**Comment by mogoodrich** [comment]
Will take a closer look at this tomorrow @mks-d  @samuelmale , thanks for the heads up!

**Pull request merged**